### PR TITLE
RFC0335: HTTP over DIDComm

### PR DESCRIPTION
**Pull request by Moopli**
Support transport of HTTP messages over a DIDComm channel.

Defines concepts, flow, and message format for a communication protocol between agents that carries HTTP messages, to enable non-agent clients/servers to use DIDComm for their communication, by interposing aries agents in between.

Signed-off-by: Filip Burlacu <filip.burlacu@securekey.com>

### CODE REVIEW

**Comment by Moopli** [comment]
All feedback is welcome.

RFC Name? Formatting? More links? Flow diagrams? Diagrammed examples? More specificity with states/flow?

**Comment by Baha-sk** [comment]
RFC looks good... probably some diagrams showing request/response flows between Aries agents would provider a better (visual) understanding.

**Comment by troyronda** [comment]
I think this could be split into two RFCS:

- Generic JSON Relay Message
  - Enable external applications to exchange JSON messages using Agent as proxy (via DIDComm connection).
  - Allows new applications to be built without changing the Agent. Particularly useful when the Agent does not need to be aware of the external application's state (nor state machine).
- HTTP Semantics on top of the JSON Relay Message (the current PR).

**Comment by troyronda** [comment]
A couple quick thoughts on a "multiplexed" JSON relay message:

- Some agents and frameworks enable an external application/service to provide the business logic (the "controller" model). 
- The JSON Relay Message extends this concept for interactions where the agent can remain stateless (beyond transient relay state). In this case, the agent simply relays for the registered application for that message.
- This approach potentially allows invocation of interoperable APIs that have been defined outside DIDComm and Aries (where the recipient agent simply relays to the external application/service). Note: the agent code does not need to be changed to support new APIs exposed by the external application/service.

**Comment by troyronda** [comment]
@dhh1128 

@Moopli has created a generic version of the JSON relay at https://github.com/hyperledger/aries-rfcs/pull/351

File: features/0335-http-over-didcomm/README.md [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+
+```jsonc
+{
+  "@type": "https://didcomm.org/http-over-didcomm/1.0/request",
+  "@id": "2a0ec6db-471d-42ed-84ee-f9544db9da4b",
+  "request": {
+    "method": "",
+    "target": "",
+    "version": "",
+    "headers": {},
```
Comment (llorllale) [inline features/0335-http-over-didcomm/README.md]: Suggest this be made into an array. [Message headers](https://tools.ietf.org/html/rfc2616#section-4.2) are allowed to appear more than once, with a little bit of parsing logic to collapse them into a single unit. It seems wiser to just let the existing, battle-hardened http components handle this.

File: features/0335-http-over-didcomm/README.md [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+
+##### `request`
+
+```jsonc
+{
+  "@type": "https://didcomm.org/http-over-didcomm/1.0/request",
+  "@id": "2a0ec6db-471d-42ed-84ee-f9544db9da4b",
+  "request": {
+    "method": "",
+    "target": "",
```
Comment (llorllale) [inline features/0335-http-over-didcomm/README.md]: Suggest rename this to "resource-uri" in order to be more aligned with the [RFC](https://tools.ietf.org/html/rfc2616#section-5.1.2)

File: features/0335-http-over-didcomm/README.md [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+
+Allows HTTP traffic to be routed over a DIDComm channel, so applications built to communicate over HTTP can make use of DID-based communication.
+
+## Motivation
+
+This protocol allows a client-server system that doesn't use DIDs or DIDComm to piggyback on top of a DID-based infrastructure, gaining the benefits of DIDs by using agents as HTTP proxies.
+
+Example use case: 
+Carl wants to apply for a car loan from his friendly neighborhood used car dealer. The dealer wants a proof of his financial stability from his bank, but he doesn't want to expose the identity of his bank, and his bank doesn't want to develop a custom in-house system (using DIDs) for anonymity. HTTP over DIDComm allows Carl to introduce his car dealer to his bank, using Aries agents and protocols, while all they need to do is install a standard agent to carry arbitrary HTTP messages.
+
```
Comment (Moopli) [inline features/0335-http-over-didcomm/README.md]: TODO: Further use cases, and further explanation of the use cases, were requested in the 18/12/19 morning call.

**Comment by swcurran** [comment]
@Moopli - are you ready for this to be merged?  There are still To Dos, as of your last comment.  If ready, right before you respond, please update the index.

**Comment by swcurran** [comment]
After reading the comments on the generic JSON RFC #351 it looks like this one should be paused. I've added the label to it. Please remove and let me know if I have misread and this is ready to merge.

**Comment by troyronda** [comment]
@Moopli Looks like there are a couple updates to this RFC with the changes to #351 .

File: features/0335-http-over-didcomm/README.md [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+DIDComm messages for this protocol look as follows:
+
+##### `request`
+
+This message has the `data-type` value `http/request`, and uses the inline `data` attachment.
+
+```
+{
+  "@type": "https://didcomm.org/https://didcomm.org/json/1.0/message",
```
Comment (troyronda) [inline features/0335-http-over-didcomm/README.md]: type has a duplicate prefix.

File: features/0335-http-over-didcomm/README.md [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+
+## Tutorial
+
+#### Name and Version
+
+This is the HTTP over DIDComm protocol. It is uniquely identified by the URI:
+
+    "https://didcomm.org/http-over-didcomm/1.0"
+    
+TODO: as a protocol defined on top of generic json messaging, does it get its own URI?
```
Comment (troyronda) [inline features/0335-http-over-didcomm/README.md]: I had thought the idea was to have a single protocol, with differing data type definitions?

**Comment by Moopli** [comment]
@swcurran we can un-pause this - all it needs is to verify links when RFC 0351 is merged.

**Comment by swcurran** [comment]
Sure - that's totally up to you. The intent is that you unpause it when you are ready.

Done.

**Comment by swcurran** [comment]
@dhh1128 - looks like this is ready for another review.

**Comment by dhh1128** [comment]
I still feel like this RFC will create confusion because of the way it special-cases HTTP 1.1 as a transport. What if I only want HTTP on *one* side of the interaction instead of both? What if I want to use the same technique with a transport like HTTP 2? We're not generalizing enough in the RFC, IMO.

However, that's not a valid reason to hold up a proposal, so I will go ahead and merge and we can work the RFC in its PROPOSED state.

**Pull request merged**